cpp/lib/src/SimpleCrawler.cc

```diff
@@ -0,0 +1,209 @@
+        if (StringUtil::StartsWith(header_line, "Location:", /* ignore_case = */ true)) {
+            std::string location(header_line.substr(9));
+            StringUtil::Trim(&location);
+            if (not location.empty())
+                location_urls->push_back(location);
+        }
+    }
+}
+
+
+void SimpleCrawler::ParseConfigFile(const std::string &config_path, std::vector<SiteDesc> * const site_descs) {
+    std::unique_ptr<File> input(FileUtil::OpenInputFileOrDie(config_path));
+
+    unsigned line_no(0);
+    while (not input->eof()) {
+        std::string line;
+        input->getline(&line);
+        ++line_no;
+        const std::string::size_type hash_pos(line.find('#'));
+        if (hash_pos != std::string::npos)
+            line = line.substr(0, hash_pos);
+        StringUtil::Trim(&line);
+        if (line.empty())
+            continue;
+
+        std::vector<std::string> line_parts;
+        if (StringUtil::SplitThenTrimWhite(line, ' ', &line_parts) != 3)
+            logger->error("in ParseConfigFile: bad input line #" + std::to_string(line_no) + " in \""
+                          + input->getPath() + "\"!");
+
+        unsigned max_crawl_depth;
+        if (not StringUtil::ToUnsigned(line_parts[1], &max_crawl_depth))
+            logger->error("in ParseConfigFile: bad input line #" + std::to_string(line_no) + " in \""
+                          + input->getPath() + "\"! (Invalid max. crawl depth: \"" + line_parts[1] + "\")");
+
+        std::string err_msg;
+        RegexMatcher * const url_regex_matcher(RegexMatcher::RegexMatcherFactory(line_parts[2], &err_msg));
+        if (url_regex_matcher == nullptr)
+            logger->error("in ParseConfigFile: bad input line #" + std::to_string(line_no) + " in \""
+                          + input->getPath() + "\", regex is faulty! (" + err_msg + ")");
+        site_descs->emplace_back(line_parts[0], max_crawl_depth, url_regex_matcher);
+    }
+}
+
+
+SimpleCrawler::SimpleCrawler(const SiteDesc &site_desc, const Params &params)
+    : remaining_crawl_depth_(site_desc.max_crawl_depth_), url_regex_matcher_(site_desc.url_regex_matcher_),
+      min_url_processing_time_(params.min_url_processing_time_), params_(params)
+{
+    // reset queues
+    std::queue<std::string> url_queue_start;
+    url_queue_start.push(site_desc.start_url_);
+    url_queue_current_depth_.swap(url_queue_start);
+    std::queue<std::string> url_queue_empty;
+    url_queue_next_depth_.swap(url_queue_empty);
+
+    // reset RegexMatchers
+    std::string err_msg;
+    url_ignore_regex_matcher_.reset(RegexMatcher::RegexMatcherFactory(params.url_ignore_pattern_, &err_msg, RegexMatcher::Option::CASE_INSENSITIVE));
+    if (url_ignore_regex_matcher_ == nullptr)
+        ERROR("could not initialize URL ignore regex matcher\n"
+              + err_msg);
+
+    // reset other class variables
+    downloader_params_                        = Downloader::Params();
```

Comment:
This line is unnecessary as the compiler has already done this at this point.  => You're doing it twice.